File: csharp/src/Yosina/Transliterators/IvsSvsBaseTransliterator.cs
// Copyright (c) Yosina. All rights reserved.

using System.Reflection;
using System.Text.Json.Serialization;
using CodePointTuple = (int First, int Second);

namespace Yosina.Transliterators;

/// <summary>
/// Transliterator for handling IVS (Ideographic Variation Sequence) and SVS (Standardized Variation Sequence) base characters.
/// </summary>
[RegisteredTransliterator("ivs-svs-base")]
public class IvsSvsBaseTransliterator : ITransliterator
{
    private readonly Options options;

    public IvsSvsBaseTransliterator()
        : this(new Options())
    {
    }

    public IvsSvsBaseTransliterator(Options options)
    {
        this.options = options ?? throw new ArgumentNullException(nameof(options));
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum Mode
    {
        [JsonStringEnumMemberName("ivs-svs")]
        IvsSvs,
        [JsonStringEnumMemberName("base")]
        Base,
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum Charset
    {
        [JsonStringEnumMemberName("unijis-90")]
        UniJis90,
        [JsonStringEnumMemberName("unijis-2004")]
        UniJis2004,
    }

    public IEnumerable<Character> Transliterate(IEnumerable<Character> input)
    {
        return this.options.Mode switch
        {
            Mode.IvsSvs => new IvsSvsBaseFwdCharEnumerator(input.GetEnumerator(), this.options),
            Mode.Base => new IvsSvsBaseRevCharEnumerator(input.GetEnumerator(), this.options),
            _ => throw new NotSupportedException($"Unsupported mode: {this.options.Mode}"),
        };
    }

    public class Options
    {
        [JsonPropertyName("mode")]
        public Mode Mode { get; set; } = Mode.IvsSvs;

        [JsonPropertyName("drop_selectors_altogether")]
        public bool DropSelectorsAltogether { get; set; }

        [JsonPropertyName("charset")]
        public Charset Charset { get; set; } = Charset.UniJis2004;

        [JsonPropertyName("prefer_svs")]
        public bool PreferSvs { get; set; }
    }

    private static class IvsSvsBaseMappings
    {
        public static readonly Lazy<IDictionary<CodePointTuple, IvsSvsBaseRecord>> FwdBase90Mappings = new(() =>
        {
            var fwdBase90Mappings = new Dictionary<CodePointTuple, IvsSvsBaseRecord>();
            foreach (var record in Mappings?.Value!)
            {
                var ivs = record.Ivs;
                var svs = record.Svs;
                var base90Tuple = record.Base90;
                var base2004Tuple = record.Base2004;

                fwdBase90Mappings[ivs] = record;
                if (!svs.IsEmpty())
                {
                    fwdBase90Mappings[svs] = record;
                }

                if (!base90Tuple.IsEmpty())
                {
                    fwdBase90Mappings[base90Tuple] = record;
                }
            }

            return fwdBase90Mappings;
        });

        public static readonly Lazy<IDictionary<CodePointTuple, IvsSvsBaseRecord>> FwdBase2004Mappings = new(() =>
        {
            var fwdBase2004Mappings = new Dictionary<CodePointTuple, IvsSvsBaseRecord>();
            foreach (var record in Mappings?.Value!)
            {
                var ivs = record.Ivs;
                var svs = record.Svs;
                var base90Tuple = record.Base90;
                var base2004Tuple = record.Base2004;

                fwdBase2004Mappings[ivs] = record;
                if (!svs.IsEmpty())
                {
                    fwdBase2004Mappings[svs] = record;
                }

                if (!base2004Tuple.IsEmpty())
                {
                    fwdBase2004Mappings[base2004Tuple] = record;
                }
            }

            return fwdBase2004Mappings;
        });

        public static readonly Lazy<IDictionary<CodePointTuple, IvsSvsBaseRecord>> RevMappings = new(() =>
        {
            var revMappings = new Dictionary<CodePointTuple, IvsSvsBaseRecord>();
            foreach (var record in Mappings?.Value!)
            {
                revMappings[record.Ivs] = record;
            }

            return revMappings;
        });

        private static readonly Lazy<IList<IvsSvsBaseRecord>> Mappings = new(() =>
        {
            var assembly = Assembly.GetExecutingAssembly();
            using var stream = assembly.GetManifestResourceStream("Yosina.Transliterators.ivs_svs_base.data");
            if (stream == null)
            {
                throw new InvalidOperationException("Could not find embedded resource ivs_svs_base.data");
            }

            using var reader = new BinaryReader(stream);
            var recordCount = reader.ReadInt32();
            if (BitConverter.IsLittleEndian)
            {
                recordCount = ReverseBytes(recordCount);
            }

            var records = new List<IvsSvsBaseRecord>(recordCount);
            for (int i = 0; i < recordCount; i++)
            {
                var ivsFirst = reader.ReadInt32();
                var ivsSecond = reader.ReadInt32();
                var svsFirst = reader.ReadInt32();
                var svsSecond = reader.ReadInt32();
                var base90 = reader.ReadInt32();
                var base2004 = reader.ReadInt32();

                if (BitConverter.IsLittleEndian)
                {
                    ivsFirst = ReverseBytes(ivsFirst);
                    ivsSecond = ReverseBytes(ivsSecond);
                    svsFirst = ReverseBytes(svsFirst);
                    svsSecond = ReverseBytes(svsSecond);
                    base90 = ReverseBytes(base90);
                    base2004 = ReverseBytes(base2004);
                }

                var ivs = ivsSecond != 0 ? (ivsFirst, ivsSecond) : (ivsFirst, -1);
                var svs = svsFirst != 0 ? (svsSecond != 0 ? (svsFirst, svsSecond) : (svsFirst, -1)) : (-1, -1);
                var base90Tuple = base90 != 0 ? (base90, -1) : (-1, -1);
                var base2004Tuple = base2004 != 0 ? (base2004, -1) : (-1, -1);

                var record = new IvsSvsBaseRecord
                {
                    Ivs = ivs,
                    Svs = svs,
                    Base90 = base90Tuple,
                    Base2004 = base2004Tuple,
                };
                records.Add(record);
            }

            return records;
        });

        private static int ReverseBytes(int value)
        {
            return (int)((value & 0x000000FFU) << 24 | (value & 0x0000FF00U) << 8 |
                         (value & 0x00FF0000U) >> 8 | (value & 0xFF000000U) >> 24);
        }
    }

    private class IvsSvsBaseFwdCharEnumerator : IEnumerable<Character>, IEnumerator<Character>
    {
        private readonly IEnumerator<Character> input;
        private readonly IDictionary<CodePointTuple, IvsSvsBaseRecord> mappings;
        private readonly bool preferSvs;
        private int offset;
        private bool disposed;

        public IvsSvsBaseFwdCharEnumerator(IEnumerator<Character> input, Options options)
        {
            this.input = input;
            this.mappings = options.Charset == Charset.UniJis90 ? IvsSvsBaseMappings.FwdBase90Mappings.Value : IvsSvsBaseMappings.FwdBase2004Mappings.Value;
            this.preferSvs = options.PreferSvs;
        }

        public Character Current { get; private set; } = null!;

        object System.Collections.IEnumerator.Current => this.Current;

        public bool MoveNext()
        {
            if (this.disposed || !this.input.MoveNext())
            {
                return false;
            }

            var c = this.input.Current;
            if (c.IsSentinel)
            {
                this.Current = c.WithOffset(this.offset);
                return true;
            }

            if (this.mappings.TryGetValue(c.CodePoint, out var replacement))
            {
                var ct = this.preferSvs && !replacement.Svs.IsEmpty() ? replacement.Svs : replacement.Ivs;
                this.Current = new Character(ct, this.offset, c);
            }
            else
            {
                this.Current = c.WithOffset(this.offset);
            }

            this.offset += this.Current.CharCount;
            return true;
        }

        public void Reset()
        {
            this.input.Reset();
            this.offset = 0;
        }

        public void Dispose()
        {
            if (!this.disposed)
            {
                this.input?.Dispose();
                this.disposed = true;
            }
        }

        public IEnumerator<Character> GetEnumerator() => this;

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => this;
    }

    private class IvsSvsBaseRecord
    {
        public CodePointTuple Ivs { get; set; }

        public CodePointTuple Svs { get; set; }

        public CodePointTuple Base90 { get; set; }

        public CodePointTuple Base2004 { get; set; }
    }

    private class IvsSvsBaseRevCharEnumerator : IEnumerable<Character>, IEnumerator<Character>
    {
        private readonly IEnumerator<Character> input;
        private readonly IDictionary<CodePointTuple, IvsSvsBaseRecord> mappings;
        private readonly Charset charset;
        private readonly bool dropSelectorsAltogether;
        private int offset;
        private bool disposed;

        public IvsSvsBaseRevCharEnumerator(IEnumerator<Character> input, Options options)
        {
            this.input = input;
            this.mappings = IvsSvsBaseMappings.RevMappings.Value;
            this.charset = options.Charset;
            this.dropSelectorsAltogether = options.DropSelectorsAltogether;
        }

        public Character Current { get; private set; } = null!;

        object System.Collections.IEnumerator.Current => this.Current;

        public bool MoveNext()
        {
            if (this.disposed || !this.input.MoveNext())
            {
                return false;
            }

            var c = this.input.Current;
            if (c.IsSentinel)
            {
                this.Current = c.WithOffset(this.offset);
                return true;
            }

            if (this.mappings.TryGetValue(c.CodePoint, out var replacement))
            {
                var ct = this.charset switch
                {
                    Charset.UniJis90 => replacement.Base90,
                    Charset.UniJis2004 => replacement.Base2004,
                    _ => (-1, -1),
                };

                if (!ct.IsEmpty())
                {
                    this.Current = new Character(ct, this.offset, c);
                    this.offset += this.Current.CharCount;
                    return true;
                }
            }

            if (this.dropSelectorsAltogether && c.CodePoint.Size() > 1)
            {
                this.Current = new Character((c.CodePoint.First, -1), this.offset, c);
            }
            else
            {
                this.Current = c.WithOffset(this.offset);
            }

            this.offset += this.Current.CharCount;
            return true;
        }

        public void Reset()
        {
            this.input.Reset();
            this.offset = 0;
        }

        public void Dispose()
        {
            if (!this.disposed)
            {
                this.input?.Dispose();
                this.disposed = true;
            }
        }

        public IEnumerator<Character> GetEnumerator() => this;

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => this;
    }
}

File: dart/lib/src/transliterators/ivs_svs_base_transliterator.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:archive/archive.dart';
import '../char.dart';
import '../transliterator.dart';
import './charset.dart';
import './ivs_svs_base.data.dart';

/// IVS/SVS base transliterator.
///
/// This transliterator handles Ideographic Variation Sequences (IVS) and
/// Standardized Variation Sequences (SVS) by either adding or removing variation selectors.
class IvsSvsBaseTransliterator implements Transliterator {
  IvsSvsBaseTransliterator({
    this.mode = 'base',
    this.charset = Charset.unijis2004,
    this.preferSvs = false,
    this.dropSelectorsAltogether = false,
  });

  static List<_IvsSvsRecord>? _records;
  static Map<Charset, Map<String, String>> revMappingsCache = {};
  static Map<(Charset, bool), Map<String, String>> fwdMappingsCache = {};

  // Options
  final String mode; // "base" or "ivs-or-svs"
  final Charset charset;
  final bool preferSvs;
  final bool dropSelectorsAltogether;

  static List<_IvsSvsRecord> _getRecords() {
    if (_records != null) return _records!;
    final bytes = ZLibDecoder().decodeBytes(base64Decode(data));
    _records = _parseData(bytes);
    return _records!;
  }

  static List<_IvsSvsRecord> _parseData(Uint8List bytes) {
    final byteData = ByteData.view(bytes.buffer);
    final recordCount = byteData.getUint32(0);
    final records = <_IvsSvsRecord>[];

    var offset = 4;
    for (var i = 0; i < recordCount; i++) {
      final ivs1 = byteData.getUint32(offset);
      final ivs2 = byteData.getUint32(offset + 4);
      final svs1 = byteData.getUint32(offset + 8);
      final svs2 = byteData.getUint32(offset + 12);
      final base90 = byteData.getUint32(offset + 16);
      final base2004 = byteData.getUint32(offset + 20);

      records.add(_IvsSvsRecord(
        ivs: String.fromCharCodes([ivs1, ivs2]),
        svs: svs1 != 0 ? String.fromCharCodes([svs1, svs2]) : null,
        base90: base90 != 0 ? String.fromCharCode(base90) : null,
        base2004: base2004 != 0 ? String.fromCharCode(base2004) : null,
      ));

      offset += 24;
    }

    return records;
  }

  Map<String, String> _getFwdMappings() {
    final key = (charset, preferSvs);
    var mappings = fwdMappingsCache[key];
    if (mappings != null) return mappings;
    mappings = <String, String>{};

    for (final record in _getRecords()) {
      final base =
          charset == Charset.unijis90 ? record.base90 : record.base2004;
      if (base != null) {
        if (record.svs != null && preferSvs) {
          mappings[base] = record.svs!;
        } else {
          mappings[base] = record.ivs;
        }
      }
    }
    fwdMappingsCache[key] = mappings;
    return mappings;
  }

  Map<String, String> _getRevMappings() {
    var mappings = revMappingsCache[charset];
    if (mappings != null) return mappings;
    mappings = <String, String>{};
    for (final record in _getRecords()) {
      final base =
          charset == Charset.unijis90 ? record.base90 : record.base2004;
      if (base != null) {
        mappings[record.ivs] = base;
        if (record.svs != null) {
          mappings[record.svs!] = base;
        }
      }
    }
    revMappingsCache[charset] = mappings;
    return mappings;
  }

  @override
  Iterable<Char> call(Iterable<Char> inputChars) sync* {
    if (mode == 'base') {
      yield* _toBase(inputChars);
    } else {
      yield* _toIvsOrSvs(inputChars);
    }
  }

  Iterable<Char> _toIvsOrSvs(Iterable<Char> inputChars) sync* {
    final mappings = _getFwdMappings();

    var offset = 0;
    for (final char in inputChars) {
      final replacement = mappings[char.c];
      if (replacement != null) {
        yield Char(replacement, offset, char);
        offset += replacement.length;
      } else {
        yield char.withOffset(offset);
        offset += char.c.length;
      }
    }
  }

  Iterable<Char> _toBase(Iterable<Char> inputChars) sync* {
    final mappings = _getRevMappings();

    var offset = 0;
    for (final char in inputChars) {
      final replacement = mappings[char.c];
      if (replacement != null) {
        // If we have a mapping, yield the base character
        yield Char(replacement, offset, char);
        offset += replacement.length;
      } else if (dropSelectorsAltogether && char.c.length > 1) {
        final c = String.fromCharCode(char.c.runes.first);
        yield Char(c, offset, char);
        offset += c.length;
      } else {
        yield char.withOffset(offset);
        offset += char.c.length;
      }
    }
  }
}

class _IvsSvsRecord {
  const _IvsSvsRecord({
    required this.ivs,
    required this.svs,
    required this.base90,
    required this.base2004,
  });
  final String ivs;
  final String? svs;
  final String? base90;
  final String? base2004;
}
